File: 002.py
#!/usr/bin/env python

from itertools import takewhile, ifilter

def fib():
    yield 1
    a,b = 1,2
    while True:
        yield b
        a,b = b,a+b

def iseven(n):
    return n % 2 == 0

def euler002():
    return sum(ifilter(iseven, takewhile(lambda x: x < 4e6, fib())))

if __name__ == "__main__":
    print euler002()

File: 002.hs
fibs = 1 : 2 : zipWith (+) fibs (tail fibs)
euler002 = sum . takeWhile (<= 4000000) . filter even $ fibs

-- alt #1
phi = (1 + sqrt 5) / 2
fib n = round $ phi^n / sqrt 5
evenFibs = map fib [3,6..]
euler002' = sum $ takeWhile (<= 4000000) evenFibs
